lib/charms/observability_libs/v1/cert_handler.py

```diff
@@ -0,0 +1,391 @@
+                logger.debug("Secret with label: 'ca-certificate-chain' not found")
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Get the certificate from the event and store it in a peer relation.
+
+        Note: assuming "limit: 1" in metadata
+        """
+        event_csr = (
+            event.certificate_signing_request.strip()
+            if event.certificate_signing_request
+            else None
+        )
+        if event_csr == self._csr:
+            content = {
+                "ca-cert": event.ca,
+                "server-cert": event.certificate,
+                "chain": json.dumps(event.chain),
+                "csr": event_csr,
+            }
+            try:
+                secret = self.model.get_secret(label="ca-certificate-chain")
+            except SecretNotFoundError:
+                if not (
+                    relation := self.charm.model.get_relation(self.certificates_relation_name)
+                ):
+                    logger.error("Relation %s not found", self.certificates_relation_name)
+                    return
+
+                secret = self.charm.unit.add_secret(content, label="ca-certificate-chain")
+                secret.grant(relation)
+                relation.data[self.charm.unit]["secret-id"] = secret.id  # pyright: ignore
+                self.on.cert_changed.emit()  # pyright: ignore
+
+    def _retrieve_from_secret(self, value: str, secret_id_name: str):
+        if not (relation := self.charm.model.get_relation(self.certificates_relation_name)):
+            return None
+
+        if not (secret_id := relation.data[self.charm.unit].get(secret_id_name)):
+            return None
+
+        if not (secret := self.model.get_secret(id=secret_id)):
+            return None
+
+        content = secret.get_content()
+        return content.get(value)
+
+    @property
+    def key(self):
+        """Return the private key."""
+        return self._private_key
+
+    @property
+    def _private_key(self) -> Optional[str]:
+        return self._retrieve_from_secret("private-key", "private-key-secret-id")
+
+    @property
+    def _csr(self) -> Optional[str]:
+        return self._retrieve_from_secret("csr", "csr-secret-id")
+
+    @_csr.setter
+    def _csr(self, value: str):
+        if not (relation := self.charm.model.get_relation(self.certificates_relation_name)):
+            return
+
+        if not (secret_id := relation.data[self.charm.unit].get("csr-secret-id", None)):
+            secret = self.charm.unit.add_secret({"csr": value})
+            secret.grant(relation)
+            relation.data[self.charm.unit]["csr-secret-id"] = secret.id  # pyright: ignore
+            return
+
+        secret = self.model.get_secret(id=secret_id)
+        secret.set_content({"csr": value})
+
+    @property
+    def _ca_cert(self) -> Optional[str]:
+        return self._retrieve_from_secret("ca-cert", "secret-id")
```

Comment:
I think `secret-id` should be a global, instead of hardcoding it here and there
Something like:

`SECRET_ID_RELATION_KEY_NAME = "secret-id"`

and then use it directly in `_retrieve_from_secret`. I think the chances that someone will override or change this key are slim.

Reply:
I can create the constant for that, but I cannot use it directly in `_retrieve_from_secret`, since for instance we have:

`self._retrieve_from_secret("csr", "csr-secret-id")` and `self._retrieve_from_secret("csr", "csr-secret-id")`